beamer/middleware.py

```diff
@@ -41,3 +51,221 @@ def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
+# the rate at which we issue requests to that RPC. In order to do that, we
+# define the following operating modes:
+#
+# 1) normal operation
+# 2) rate limiting mode
+# 3) tapering mode
+#
+# In normal operation, the RPC accepts our every request and we issue requests
+# at full rate, i.e. as they come.
+#
+# Rate limiting mode begins when the RPC returns HTTP 429, while we are in
+# normal operation. For the duration of this mode, defined by _RATE_LIMIT_PERIOD,
+# we will attempt to issue requests with a reduced rate, defined by
+# _RATE_LIMIT_REQUEST_DELAY, the number of seconds between each attempt.
+#
+# This effectively caps the request rate at 1 / _RATE_LIMIT_REQUEST_DELAY with
+# the hope that the RPC will eventually notice that we slowed down and accept
+# the requests. If the RPC rejects new requests even after the rate limiting
+# period is over, we simply raise a RuntimeError, which will cause the agent to
+# shut down, as there is nothing more that could be done.
+#
+# Tapering mode begins immediately after the end of rate limiting mode, with
+# the idea that we ramp up to the original normal operation request rate (i.e.
+# without any delays). Currently,t the ramp up is implemented in the following
+# way:
+#
+# - upon exiting the rate limiting mode, set
+#   - _RateLimiterState.taper_counter to 0
+#   - _RateLimiterState.taper_counter_max to the number of threads currently
+#    waiting on the rate limiter lock in the _rate_limiter function
+#
+# - as each thread that waited on the lock gets its turn, it will notice it is
+#   in the tapering mode (taper_counter_max > 0) so it will sleep for a delay of
+#
+#       _RATE_LIMIT_REQUEST_DELAY / 2**state.taper_counter
+#
+# - when taper_counter reaches taper_counter_max, the tapering mode is over and
+#   we return to normal operation
+#
+# - if at any point during tapering mode we get rate limited by the RPC, we
+#   will immediately switch back to rate limiting mode
+#
+# The tapering mode tries to quickly ramp up to normal request rate by
+# exponentially reducing the delay time between consecutive requests.
+
+
+# The rate limiter state.
+#
+# One instance per web3 rate limiter middleware, i.e. one instance per
+# chain/RPC if there is exactly one Web3 instance per chain/RPC.
+#
+# Note: it is important to ensure only one Web3 instance per chain/RPC,
+# otherwise, it could happen that two parallel rate limiter middleware talk to
+# the same RPC, which would rate limit both middlewares.
+#
+# The lock protects all state fields except num_waiting_on_lock, which is an
+# atomic used to keep track of threads waiting on the lock.
+@dataclass(slots=True)
+class _RateLimiterState:
+    lock: threading.Lock = field(default_factory=threading.Lock)
+    num_waiting_on_lock: atomics.base.AtomicUint = field(
+        default_factory=lambda: atomics.atomic(width=4, atype=atomics.UINT)
+    )
+    rate_limit_end: None | float = None
```

Comment:
isnt this equal to Optional[float] ?

Reply:
Indeed it is, this is by the new union types feature in newer Python: https://docs.python.org/3/library/stdtypes.html#types-union so `Optional` isn't really needed anymore.